content/blog/2025-10-02-50-million-zsets/index.md

```diff
@@ -0,0 +1,65 @@
+date=2025-10-02 00:00:01
+description= """ 
+    hi
+    """
+authors= ["khawaja"]
+[extra]
+featured = true
+featured_image = "/assets/media/featured/random-08.webp"
++++
+
+When you run infrastructure at scale, the smallest efficiencies compound into massive savings. Sorted sets (ZSETs) are the backing data structure for far more than leaderboards. They're used for time‑ordered feeds, priority queues, recommendation rankings and more. Each entry carries per‑key overhead; when you're inserting tens of millions of items, those bytes accumulate into gigabytes. The latest Valkey 8.1 release introduces a [redesigned hash table](https://valkey.io/blog/valkey-8-1-0-ga/) and other optimizations that promise lower memory usage and higher throughput. In this post, we put Valkey 8.1 under pressure by benchmarking it against Valkey 8.0, inserting 50 million members into a sorted set and measuring memory consumption and throughput along the way.
+
+## About Momento and Raider.IO
+
+Momento builds enterprise‑grade real‑time data infrastructure. Our flagship service, [Momento Cache](https://gomomento.com), allows developers to focus on innovation rather than maintaining complex back‑ends. Behind the curtains, Valkey is one of the primary storage engines we operate. Its memory efficiency and predictable performance make it an ideal choice for mission‑critical workloads. By building on Valkey, we can deliver caching experiences that are fast, reliable, and cost‑effective.
+
+One such workload is [Raider.IO](https://raider.io), the most popular companion site for World of Warcraft players. Raider.IO maintains leaderboards that track millions of characters and guilds, updating instantly whenever a player clears a dungeon or raid. That translates into hundreds of millions of sorted set operations every day. In this environment, downtime or latency isn't acceptable. Memory efficiency and throughput directly shape the player experience. Running Valkey at hyperscale for workloads like Raider.IO means that improvements in Valkey 8.1 immediately impact both infrastructure cost and the ability to absorb peak traffic without breaking a sweat.
+
+## Benchmark Setup
+
+To provide a fair comparison, both Valkey 8.0 and Valkey 8.1.1 were run on the same hardware:
+
+* AWS c8g.2xl instance (Graviton4, 8 vCPU, 16 GB RAM)
+* Identical settings for persistence, I/O threads and memory limits
+* 50 million members inserted into a ZSET via pipelined ZADD commands (score=i, member=m:{i})
+* 250,000 item batch size per pipeline
+* Metrics collected every 1,000,000 inserts (flush between runs)
+
+The benchmark code is open sourced and straightforward to reproduce: it connects to both servers, flushes the test key, performs batched inserts, and records `used_memory`, `used_memory_rss`, total elapsed time, and throughput after each million inserts. This repeatability mirrors the ethos of Valkey's community - every optimization is measurable, and anyone can verify the results.
```

Comment:
Maybe also include the link to the open source benchmark here, so you don't need to wonder where if someone doesn't finish the blog.

Reply:
Ah good call, here's the link to it in GitHub

> https://github.com/momentohq/sorted-set-benchmark